core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/batch_limiter_middleware.rs

```diff
@@ -0,0 +1,96 @@
+use std::{num::NonZeroU32, sync::Arc};
+
+use governor::{
+    clock::DefaultClock,
+    middleware::NoOpMiddleware,
+    state::{InMemoryState, NotKeyed},
+    Quota, RateLimiter,
+};
+use vise::{Buckets, Counter, EncodeLabelSet, EncodeLabelValue, Family, Histogram, Metrics};
+use zksync_web3_decl::jsonrpsee::{
+    server::middleware::rpc::{layer::ResponseFuture, RpcServiceT},
+    types::{error::ErrorCode, ErrorObject, Request},
+    MethodResponse,
+};
+
+use crate::api_server::web3::metrics::API_METRICS;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "transport", rename_all = "snake_case")]
+pub(crate) enum Transport {
+    Ws,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "api_jsonrpc_backend_batch")]
+struct LimitMiddlewareMetrics {
+    /// Number of rate-limited requests.
+    rate_limited: Family<Transport, Counter>,
+    /// Size of batch requests.
+    #[metrics(buckets = Buckets::exponential(1.0..=512.0, 2.0))]
+    size: Family<Transport, Histogram<usize>>,
+    /// Number of requests rejected by the limiter.
+    rejected: Family<Transport, Counter>,
+}
+
+#[vise::register]
+static METRICS: vise::Global<LimitMiddlewareMetrics> = vise::Global::new();
+pub struct LimitMiddleware<S> {
```

Comment:
- Nit: Missing an empty line between items.
- Nit: Visibility could be reduced to `pub(crate)`.
- Nit: Could make sense to add a doc comment that an instance of this middleware is allocated per connection, so that it would be easier to reason about it for people not very familiar with `jsonrpsee`.

Reply:
fixed.